src/docker.ts

```diff
@@ -26,13 +36,32 @@ const saveDockerImages = async (): Promise<void> => {
       `Cache miss occurred on the primary key ${key}. Not saving cache as read-only option was selected.`
     );
   } else {
-    await execBashCommand(
-      'docker image list --format "{{ .Repository }}:{{ .Tag }}" | ' +
-        '2>&1 xargs --delimiter="\n" --no-run-if-empty --verbose --exit ' +
-        `docker save --output ${DOCKER_IMAGES_PATH}`
+    const unneededDockerImagesList = getState(DOCKER_IMAGES_LIST).split("\n");
+    const fullDockerImagesList = (
+      await execBashCommand(
+        'docker image list --format "{{ .Repository }}:{{ .Tag }}"'
+      )
+    ).split("\n");
+    console.log(`fullDockerImagesList = ${fullDockerImagesList}`);
+    const dockerImagesToBeSaved = fullDockerImagesList.filter(
+      (image) => !unneededDockerImagesList.includes(image)
     );
+    if (dockerImagesToBeSaved.length === 0) {
+      info("No docker images to save");
+    } else {
+      const formattedDockerImagesToBeSaved = dockerImagesToBeSaved.join(" ");
+      await execBashCommand(
+        `docker save --output ${DOCKER_IMAGES_PATH} ${formattedDockerImagesToBeSaved}`
+      );
+    }
     await saveCache([DOCKER_IMAGES_PATH], key);
```

Comment:
Why are we calling `saveCache` if there are no Docker images to save? This is a good opportunity for test-driven development: improve the tests so that the current production code fails, then correct the production code so that the improved tests pass.

Reply:
Oh wow, good point.  I didn't notice that!!

Reply:
I'm failing to find a reasonable way to implement this.  None of the current test cases in `docker.ts` seem to me to be obvious matches for the scenario where there are no Docker images to save.  Thus, I looked in the `integration.test.ts` and figured that it could go in the cache miss part of the integration test.  However, then I realized that with the current mocking situation, `child_process.exec.mockImplementation` is designed to increment the counter in the case where the command is `docker image list...` so that we never will get a situation where we are recreating a simulation of no Docker images to save.  I feel like I could maybe make a separate branch for the mock but this seems messy.  Am I missing something obvious/what is a better way of thinking of this??  Thanks in advance for your thoughts.

Reply:
Oops, you are right. We need a new test case in `docker.test.ts` for this scenario.